sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/SubscriptionPartitionLoader.java

```diff
@@ -47,15 +46,99 @@
   private final SubscriptionPath subscription;
   private final SerializableFunction<TopicPath, Integer> getPartitionCount;
   private final Duration pollDuration;
-  private final boolean terminate;
+  private final SerializableSupplier<Boolean> terminate;
+
+  private class GeneratorFn extends DoFn<Void, SubscriptionPartition> {
+    @ProcessElement
+    public ProcessContinuation processElement(
+        OutputReceiver<SubscriptionPartition> output,
+        RestrictionTracker<Integer, Integer> restrictionTracker) {
+      int previousCount = restrictionTracker.currentRestriction();
+      int newCount = getPartitionCount.apply(topic);
+      if (newCount <= previousCount) {
+        return ProcessContinuation.resume().withResumeDelay(pollDuration);
+      }
+      if (!restrictionTracker.tryClaim(newCount)) {
+        return ProcessContinuation.stop();
+      }
+      Instant ts = previousCount == 0 ? BoundedWindow.TIMESTAMP_MIN_VALUE : getWatermark();
+      for (int i = previousCount; i < newCount; ++i) {
+        output.outputWithTimestamp(SubscriptionPartition.of(subscription, Partition.of(i)), ts);
+      }
+      if (terminate.get()) {
+        return ProcessContinuation.stop();
+      }
+      return ProcessContinuation.resume().withResumeDelay(pollDuration);
```

Comment:
Done with modifications- in particular, we still need to advance the watermark if there is no new data to prevent UI timestamps from growing without bound for this.